### ISSUE TITLE
Clarify documentation

### ISSUE THREAD

**Issue by charlesreid1** [mkdocs/mkdocs#1452]
Hello,

First, mkdocs is a great tool and I've been really enjoying it, so thanks to the developers.

The [Customizing a Theme](http://www.mkdocs.org/user-guide/styling-your-docs/#customizing-a-theme) section of the documentation could use some clarification. I have spent about an hour tinkering with it, but still can't understand how it works.

* If I want to add an external CSS file, docs tell me I can use the `docs_dir`. Okay.
* Docs give me a simple example CSS file that I can drop into `docs_dir/extras.css`. Okay.
* Note in the docs says, if I'm deploying with readthedocs (which I'm not), I need to include the CSS with the `extra_css: [extra.css]` directive explicitly in my YAML file. I'm not using readthedocs, so I don't.

When I serve the documentation with `mkdocs serve`, the CSS file is not picked up. The CSS file is not included in the HTML at all. So, **the blue note mentioning readthedocs is incorrect or not clear enough**.

The documentation picks up after that: "After making these changes, they should be visible when you run mkdocs serve" - but **it's unclear what changes the documentation is referring to**. The `extra_css` change mentioned in the note? Or creating the `extras.css` CSS file?

I ended up having to add the `extra_css: [extra.css]` bit into my YAML configuration file to get it to add `extra.css` to my index, so the blue note mentioning readthedocs apparently applies to **all** cases.

However, even then it took more tinkering, because `extra_css: [extra.css]` was adding a link to `extra.css` in the current directory, not the `css/` directory where it was actually located.

I could keep tinkering, e.g., by changing `extra_css: [css/extra.css]` (which I think is valid), but that leads to another problem, which is that `extra.css` is not being copied into the `css/` directory when I create the site.

Without better documentation, I'm stuck with banging my head against a wall... or just picking a different theme.

I've been really happy with mkdocs, all the other features are pretty straightfoward. It's just this feature I'm having trouble with, mainly because of vague/confusing documentation.

**Comment by waylan** [comment]
Thanks for pointing this out. As the [release notes][1] mention, the behavior of the `extra_css` setting has recently changed. It appears that we updated the [configuration][2] docs, but not the Theme Customization docs.

> I could keep tinkering, e.g., by changing extra_css: [`css/extra.css`] (which I think is valid), but that leads to another problem, which is that `extra.css` is not being copied into the `css/` directory when I create the site.

Would it help to understand that all files and directories in the `docs_dir` maintain their structure when being copied to the `site_dir`? If your file is at `extra.css` in the `docs_dir`, then it needs to be `extra_css: ['extra.css']` in the config and will be at the location `extra.css` in the `site_dir`. Conversely, if our file is at `css/extra.css` in the `docs_dir`, then it needs to be `extra_css: ['css/extra.css']` in the config and will be at the location `css/extra.css` in the `site_dir`.

> Docs give me a simple example CSS file that I can drop into `docs_dir/extras.css`.

In case it's not clear, `docs_dir` should be the value of the `docs_dir` settings (`docs` by default), not literally `"docs_dir"` (unless you have `docs_dir: docs_dir` in your config). In other words the same directory where your homepage `index.md` would go.

If you are still having trouble, please post your config and the listing of directories and files in your `docs_dir` and we'll help you work it out.

[1]: http://www.mkdocs.org/about/release-notes/#auto-populated-extra_css-and-extra_javascript-fully-deprecated-986
[2]: http://www.mkdocs.org/user-guide/configuration/#extra_css

**Comment by charlesreid1** [comment]
> In case it's not clear, docs_dir should be the value of the docs_dir settings (docs by default), not literally "docs_dir"

d'oh! I think this misunderstanding was the central problem. The mkdocs behavior started to make more sense after I read that. I'll put together a PR for this particular section of the docs...

**Comment by charlesreid1** [comment]
Requesting more clarification in the documentation. This time it's around themes.

The documentation does not clearly cover how to install a theme without pypi. For example, I want to modify (majorly modify) a theme that is pypi-installable, so I clone a copy of it from Github. Once I make modifications to the theme, and I want to use it on multiple sites, it isn't clear how I proceed.

The documentation covers the cases:

* Adding custom files to the `docs_dir` (as per above) to modify an existing
* Pip installing an existing theme

But it does not cover the case of "I have my own custom mkdocs theme in some arbitrary directory, how do I use it?"

The documentation covers how to create themes/template files just fine, but when it comes time to install it, pointing to your own arbitrary theme directory is not covered on the [Customize Theme](http://www.mkdocs.org/user-guide/custom-themes/) page or on the [Styling Your Docs](http://www.mkdocs.org/user-guide/styling-your-docs/) page as far as I can see.

There should at least be an "Install" section on one of these pages, to specify how you install themes. I imagine this is some simple command like `mkdocs install-theme mkdocs-awesome-theme`, but I can't find what I'm looking for.

**Comment by waylan** [comment]
> But it does not cover the case of "I have my own custom mkdocs theme in some arbitrary directory, how do I use it?"

This is covered in [Creating a custom theme][1]. No "installation" is necessary. Set `theme.name` to `null` and then point `theme.custom_dir` to your arbitrary directory. In the example provided, the arbitrary directory is named `custom_theme`. 

If you want to "install" a theme, first you have to wrap it in a package for distribution, which is covered in [Packaging Themes][2]. Once the theme is wrapped in a package, then standard Python packaging applies and our docs link to the standard Python packaging docs in multiple locations.

[2]: https://www.mkdocs.org/user-guide/custom-themes/#packaging-themes
[1]: https://www.mkdocs.org/user-guide/custom-themes/#creating-a-custom-theme

**Comment by charlesreid1** [comment]
Ok thanks, will roll this into the PR. Will note that a lot of the confusion is boiling down to the YAML examples, specifically YAML that point to folders/directories, without that fact being clear. If a path can be absolute, it would be good to use something like `/path/to/whatev` instead of `whatev`, or otherwise add a trailing slash `whatev/`. Or heck, be wordy and give an example of both!

**Comment by waylan** [comment]
Thanks for the feedback. This is often the problem when writing documentation. When you know how it works it can be easy to leave stuff out that seems obvious to you. Sometimes it takes a newcomer to point out the missing pieces.

**Comment by waylan** [comment]
While working on #1376 I updated the documentation referenced here. Any lack of clarity was addressed, although I noted that following the links to the `configuration` page makes it clear that each of the relevant settings are to point to a directory. The problems described above seem to be related to a failure to reference the descriptions of the [configuration options][1].

 [1]: https://www.mkdocs.org/user-guide/configuration/

**Comment by charlesreid1** [comment]
Addressed in #1463

**Comment by charlesreid1** [comment]
Going to point out another inconsistency between the tool and the docs (or, between your statement): you said there is no "installation" of themes necessary. However, I'm currently trying to use a custom theme and the tool cannot pick up the configuration file for it. The error message states:

```
$ mkdocs serve
INFO    -  Building documentation...
ERROR   -  Config value: 'theme'. Error: Unrecognised theme name: 'stupid'. The available installed themes are: mkdocs, readthedocs, alabaster
```

Why/how is it listing available "installed" themes if themes are not installed? This is confusing because even though what you are saying (about themes not needing to be installed - everything's flat) makes sense, it doesn't match the behavior I am seeing when I use mkdocs (i.e., I cannot for the life of me get it to use a theme that has not been "pip install"ed or that is not included with mkdocs stock).

The documentation and your explanations are helping, but this is all still very unclear.

It would be helpful if there were a single documentation page collecting information about customizing themes (using a built-in theme, and where they live; using a built-in theme, but adding custom JS or CSS; using someone else's theme via "pip install mkdocs-theme"; and defining your own theme from the ground up). Currently this info is spread across multiple pages, so even finding information I know is there is difficult.

**Comment by charlesreid1** [comment]
Okay, I think I understand - if I specify `name: something` then it means I've run `pip install mkdocs-something`, so I have to say `name: null` if I want to use _local_ theme files.

**Comment by Alex-Cretu** [comment]
I have a similar problem.
I used the `custom_dir: folder_name` setting, in `folder_name` I had the theme file structure as described in the documentation (Used the files from the `mkdocs` theme).

When I execute `mkdocs serve` I get this: 

```
INFO    -  Building documentation...
INFO    -  Cleaning site directory
ERROR   -  Error building page index.md
Traceback (most recent call last):
  File "/usr/local/bin/mkdocs", line 11, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/mkdocs/__main__.py", line 127, in serve_command
    livereload=livereload
  File "/usr/local/lib/python2.7/site-packages/mkdocs/commands/serve.py", line 117, in serve
    config = builder()
  File "/usr/local/lib/python2.7/site-packages/mkdocs/commands/serve.py", line 112, in builder
    build(config, live_server=live_server, dirty=dirty)
  File "/usr/local/lib/python2.7/site-packages/mkdocs/commands/build.py", line 282, in build
    build_pages(config, dirty=dirty)
  File "/usr/local/lib/python2.7/site-packages/mkdocs/commands/build.py", line 242, in build_pages
    _build_page(page, config, site_navigation, env)
  File "/usr/local/lib/python2.7/site-packages/mkdocs/commands/build.py", line 150, in _build_page
    template = env.get_template('main.html')
  File "/usr/local/lib/python2.7/site-packages/jinja2/environment.py", line 830, in get_template
    return self._load_template(name, self.make_globals(globals))
  File "/usr/local/lib/python2.7/site-packages/jinja2/environment.py", line 804, in _load_template
    template = self.loader.load(self, name, globals)
  File "/usr/local/lib/python2.7/site-packages/jinja2/loaders.py", line 125, in load
    code = environment.compile(source, name, filename)
  File "/usr/local/lib/python2.7/site-packages/jinja2/environment.py", line 591, in compile
    self.handle_exception(exc_info, source_hint=source_hint)
  File "/usr/local/lib/python2.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/alexandrucretu/test-project/xldocs/main.html", line 20, in <module>
    <link href="{{ 'css/bootstrap-custom.min.css'|url }}" rel="stylesheet">
jinja2.exceptions.TemplateAssertionError: no filter named 'url'
```

Is it a problem with the `custom_dir` or `pip` or what do you think?
Thanks

**Comment by waylan** [comment]
@Alex-Cretu you appear to have an unrelated problem. Your config is fine and MkDocs is finding your templates. The problem is with the template itself. It would seem that you are using a template designed for MkDocs version 1.0, however, you are using an older version of MkDocs which does not yet support some of the features adding in version 1.0. Please follow up in a new issue.

**Comment by Alex-Cretu** [comment]
@waylan Thank you for the fast reply. I used the 0.17 theme and modified it and in works. 

I have an unrelated question. We want to migrate our project from a Jekyll based site to mkdocs. As I can see, if the `md` files are not specified under `pages` in  `mkdocs.yml` they can't be accessed even if they are placed in `docs` dir. Is there any other solution for this? We have hundreds of files in nested folders.

**Comment by waylan** [comment]
This will be addressed in version 1.0, which is currently available as a beta release (1.0b1). Also your 1.0 templates would work with this as well. Preview the [release notes][1] for details.

And let's stop adding noise to the discussion about this unrelated issue. Please direct support questions to the [mailing list][2] or open a new issue to report a bug or make a feature request. Thank you.

[1]: https://github.com/mkdocs/mkdocs/blob/master/docs/about/release-notes.md#version-10-2018--
[2]: https://groups.google.com/forum/#!forum/mkdocs
